src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -242,6 +266,18 @@ static NpgsqlException NoSuitableHostsException(IList<Exception> exceptions)
                 : new("Unable to connect to a suitable host. Check inner exception for more details.",
                     new AggregateException(exceptions));
 
+        int GetRoundRobinIndex()
+        {
+            lock (_roundRobinIndexLock)
```

Comment:
Yeah, that actually was my initial implementation.

```csharp
var value = Interlocked.Increment(ref this._counter) % this.MaxCounterValue;
if (value == 0)
{
   Interlocked.Add(ref this._counter, -this.MaxCounterValue);
}

return value;
```

There are a few problems here:
1. It's a little bit harder to understand
2. It always skips the first host the first time it's called

Reply:
> Is the value == 0 for wraparound handling? Would the value be negative?

Yep.

Reply:
But then checking `value == 0` wouldn't work, right? Because _counter would be int.MinValue and not 0?

Reply:
It would work up to a point. -1 is incremented to 0, then 0 % 0 returns 0, then it's decremented to some negative number, and after that it will fail while attempting to get a pool with a negative index.

Reply:
I'm confused... I was thinking about the following:

1. Initialize _counter to -1
2. Use Interlocked.Increment
3. If it's less than int.MaxValue, just modulu by the number of pools and return that
4. If we've reached int.MaxValue, we need to wraparound, this can be done by doing CompareAndExchange to try to set to zero. If succeeded, return zero. If not, someone else already did it (concurrency) so we can try doing a simple Interlocked.Increment again.

It's indeed a bit more complex than a simple lock, and I admit I'm not sure there's a huge benefit over locking, but it still feels like the right thing to do... What do you think?

Reply:
> If we've reached int.MaxValue, we need to wraparound, this can be done by doing CompareAndExchange to try to set to zero. If succeeded, return zero. If not, someone else already did it (concurrency) so we can try doing a simple Interlocked.Increment again.

That's not going to work - if you increment int.MaxValue it just wraps around, so it will be a race condition between some other thread incrementing the counter, and the CompareAndExchange resetting it to 0.

Reply:
Right, the above isn't the exact code. If the result of Interlocked.Increment is negative, try setting to zero with CompareAndExchange. If you failed, retry the whole thing again - someone else already set it to zero.

If you want I can write the code?

Reply:
> Right, the above isn't the exact code. If the result of Interlocked.Increment is negative, try setting to zero with CompareAndExchange. If you failed, retry the whole thing again - someone else already set it to zero.

Or incremented it...